### ISSUE TITLE
CELLULOID_DEBUG blocks fork() remedy.

### ISSUE THREAD

**Issue by digitalextremist** [celluloid/celluloid#517]
18a125e374d73a69ae78a27b497fedebd497a5b5 causes tests to fail on the fork() remedy: #515

See notes on commit.

/cc: @tarcieri, @e2

**Comment by e2** [comment]
@digitalextremist - way too much confusion here. Sorry if I added to it.

Just to clear things up:
1 @tarcieri broke tests on master with CELLULOID_DEBUG thing
2. Meanwhile, I created a simple patch for internal_pool.rb for handling fork()
3. I created a PR on master (not knowing master had 10 failing tests)
4. you merged before the tests finished (I didn't expect master to have failing tests)
5. there were actually 10 failing tests - I found the cause was the CELLULOID_DEBUG commit
6. you asked me for a PR on the stable branch
8. I created a PR for stable (since the bad commit is not on stable, merging this would be fine - especially since tests pass)
8. You put the stable merge on hold (no reason technically)
9. You reverted my PR on master (no reason technically) instead of reverting the CELLULOID_DEBUG commit (which would have resulted in passing tests - including those for my patch for fork())

What you could have been simpler (if you somehow knew exactly what the patches did, why the tests failed, etc.):
1. Given you merged my patch into master ...
2. ... you could have just reverted tarcieri's test-breaking commit on top of master
3. you could have just merged master into 0-16-stable (since tests would have passed)

There's so much going on, I don't want to add unnecessary work for you - just tag me if you need me.

**Comment by digitalextremist** [comment]
@e2 I held the PR(s) based on the confusion, not based on the validity of the code. So I appreciate the play-by-play.

Even if/since/when the 18a125e withdrawal fixes it all, I'd rather the code sit a little bit until there is zero confusion. I merged the PR because I am trying to get `0.16.5` landed before GSoC, and there is still a lot of time for that. I understand the patch, I mostly understand the breakage, but like I said, confusion is a good reason to revert, rather than carry forward. I don't doubt the patch is good.

**Comment by e2** [comment]
@digitalextremist - it's not about the patch, it's about the confusion that could have been avoided if the PR sat there until the tests finished (for quality). 

It's my fault for not running all the specs locally (the InternalPool specs were fine, and I didn't touch anything else, so I was pretty sure Travis would pass).

It's your fault for accepting the patch with unfinished tests ;)

Also, there are only 5 non-merge commits between 0-16-stable and master (one including the revert) - none of them affect Celluloid's stability, really (except for broken tests, which rely on debugging being enabled).

I could've just created a patch to fix the bad commit ... but I thought that would be more confusing than just asking to revert it.

If you need to be "notified" when tests succeed on a patch of mine, just let me know (and I'll mention you saying it's ready for merging).

If a patch doesn't result in green tests on master, it's a bad patch - even if that patch isn't the one that caused the problem (and so, it shouldn't be merged to master in the first place).

Let me know what I can / could have done to resolve this as a simple+single PR in the future - thanks.

**Comment by digitalextremist** [comment]
@e2 haha, it's really simple actually -- don't tell me the merge is ready until it's ready :) Much less for `-`*-stable`! After all the bullet-point lists outlining protocol recently and all your comments about making sure tests pass I was *sure* you followed all that before you called for a merge :)

**Comment by e2** [comment]
@digitalextremist - for me "it's ready if Travis has no objections". If Travis has objections, it's not ready.

I don't trust my local tests, because I may e.g. forget to add a file, or push/pull the wrong branch, my master may be behind (it's why Travis runs 2 sets of tests), or ... I could have a race condition that locally accidentally results in succeeding tests (not unlikely in Cellluloid).

I don't trust myself - which is why I depend on unit tests and Travis also.
The PR *was* ready for stable - it's just master that was broken "underneath" the PR.

In short: my PR was ready for `master`, but `master` wasn't ready for my patch.

What's the point of having Travis if you're skipping it? If master can be broken, what's the point of reverting changes? I just can't seem to explain that to myself in my head.

Also, you can always break master between my PR and merging it - I have no control over that.

I don't want this to break into another discussion - so let's just say: "don't merge until I say it's ready" - especially if Travis hasn't finished or isn't green. Or do you want me to say "please wait for Travis" in every PR I make?

**Comment by digitalextremist** [comment]
My joke backfired @e2! Suffice to say *I understand* ...

**Comment by e2** [comment]
@digitalextremist - I honestly don't know when you're joking.

**Comment by tarcieri** [comment]
@e2 I ran the tests locally. There are many broken, flaky tests.

We need to get rid of all of them. That's the only solution.

Would you like to help with that or would you just like to complain?

**Comment by e2** [comment]
@tarcieri - 10 tests were broken by your commit (@18a125e). NOT flaky tests, but DETERMINISTIC tests. 

Specifically, tests requiring CELLULOID_DEBUG to be true to run, because they were testing for debug messages.

Those 10 failed on Travis and they failed locally for me. I did a revert locally on that commit and then all tests were fine - both with and without my patch.

There's no merge commit -your commit (@18a125e) was directly committed to master. No big deal for me - I mentioned it only to explain why there were Travis failures in my tiny PR.

I don't have commit access directly to master, but if I did, I'd simply revert that commit without a PR. That's what I expected you or @digitalextremist or any other maintainer of ANY project in this situation to do.

Instead, my own patch was merged and then reverted.

I have NO INFLUENCE on this:
1. I couldn't prevent my PR from being merged with a Travis check
2. The bad commit wasn't reverted
3. My own commit was reverted (SERIOUSLY, WTF?)
4. I created a backport PR, which was put on hold (for no technical reason - all tests pass)
5. All I get is BS or irrelevant comments having nothing to do with the facts

Flaky tests are not the problem here - the problem is a complete misunderstanding of what's going on.

This is the most NONSENSE-FILLED situation I've ever experienced in history! This is worse than Kafka's "Process".

How am I supposed to help if a tiny, non-intrusive, test-covered, working PR gets merged without tests completing, then gets reverted for no reason, then you yourself bring up something as irrelevant here as flaky unit tests?

I'm just COMPLETELY BAFFLED! Seriously, @digitalextremist and @tarcieri ...

... WHAT ... THE ... F*CK is going on here?

**Comment by e2** [comment]
@tarcieri - sorry, I just saw your revert on master.

Issue is closed for me. I was beginning to think I'm going crazy.

**Comment by tarcieri** [comment]
@e2 here's a branch with the change:

https://github.com/celluloid/celluloid/pull/518

Care to actually help figure out why this is making the tests fail?

**Comment by digitalextremist** [comment]
This was resolved, and is being backported in #523.
